mail/common/src/main/java/com/fsck/k9/mail/oauth/AndroidAccountOAuth2TokenStore.java

```diff
@@ -0,0 +1,89 @@
+import android.accounts.AccountManagerFuture;
+import android.app.Activity;
+import android.content.Context;
+import android.os.Bundle;
+
+import com.fsck.k9.mail.AuthenticationFailedException;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+;
+
+/**
+ * An interface between the OAuth2 requirements used for authentication and the AccountManager.
+ */
+public class AndroidAccountOAuth2TokenStore implements OAuth2TokenProvider {
+    private static final String GMAIL_AUTH_TOKEN_TYPE = "oauth2:https://mail.google.com/";
+    private static final String GOOGLE_ACCOUNT_TYPE = "com.google";
+
+    private Map<String,String> authTokens = new HashMap<>();
+    private AccountManager accountManager;
+    private Context context;
+
+    public AndroidAccountOAuth2TokenStore(Context applicationContext) {
+        this.accountManager = AccountManager.get(applicationContext);
+        this.context = context;
+    }
+
+    @Override
+    public void authorizeApi(final String emailAddress, final Activity activity, final OAuth2TokenProviderAuthCallback callback) {
+        Account account = new Account(emailAddress, GOOGLE_ACCOUNT_TYPE);
+        accountManager.getAuthToken(account, GMAIL_AUTH_TOKEN_TYPE, null,
+                activity, new AccountManagerCallback<Bundle>() {
+                    @Override
+                    public void run(AccountManagerFuture<Bundle> accountManagerFuture) {
+                        try {
+                            accountManagerFuture.getResult();
+                            callback.success();
+                        } catch (Exception e) {
+                            callback.failure(new AuthorizationException("Authorization failed", e));
+                        }
+                    }
+                }, null);
+    }
+
+    @Override
+    public String getToken(String username, long timeoutMillis) throws AuthenticationFailedException {
+        if(authTokens.get(username) == null) {
+            fetchNewAuthToken(username, timeoutMillis);
+        }
+        return authTokens.get(username);
+    }
+
+    private void fetchNewAuthToken(String username, long timeoutMillis)
+            throws AuthenticationFailedException {
+        try {
+            Account account = new Account(username, GOOGLE_ACCOUNT_TYPE);
+            AccountManagerFuture<Bundle> future = accountManager
+                    .getAuthToken(account, GMAIL_AUTH_TOKEN_TYPE, false, null, null);
+            Bundle bundle = future.getResult(timeoutMillis, TimeUnit.MILLISECONDS);
+            if (bundle == null) {
+                throw new AuthenticationFailedException("No token provided");
+            }
+            if (bundle.get(AccountManager.KEY_ACCOUNT_NAME) == null) {
+                throw new AuthenticationFailedException("No account information provided");
+            }
+            if (bundle.get(AccountManager.KEY_ACCOUNT_NAME).equals(username)) {
```

Comment:
I'm not entirely sure that these checks are important, if we ask for account `X` the result being for account `Y` seems unlikely...

Reply:
Paranoia from seeing what some Android manufacturers manage to ship.